src/request/plan.rs

```diff
@@ -63,42 +66,213 @@ impl<Req: KvRequest + HasKeys> HasKeys for Dispatch<Req> {
+    // 2. Ok(false): backoff, and then retry
+    // 3. Err(Error): can't be resolved, return the error to upper level
+    async fn handle_region_error(
+        pd_client: Arc<PdC>,
+        mut e: tikv_client_proto::errorpb::Error,
+        region_store: RegionStore,
+    ) -> Result<bool> {
+        let ver_id = region_store.region_with_leader.ver_id();
+        if e.has_not_leader() {
+            let not_leader = e.get_not_leader();
+            if not_leader.has_leader() {
+                match pd_client
+                    .update_leader(
+                        region_store.region_with_leader.ver_id(),
+                        not_leader.get_leader().clone(),
+                    )
+                    .await
+                {
+                    Ok(_) => Ok(true),
+                    Err(e) => {
+                        pd_client.invalidate_region_cache(ver_id).await;
+                        Err(e)
+                    }
+                }
+            } else {
+                // The peer doesn't know who is the current leader. Generally it's because
+                // the Raft group is in an election, but it's possible that the peer is
+                // isolated and removed from the Raft group. So it's necessary to reload
+                // the region from PD.
+                pd_client.invalidate_region_cache(ver_id).await;
+                Ok(false)
+            }
+        } else if e.has_store_not_match() {
+            pd_client.invalidate_region_cache(ver_id).await;
+            Ok(false)
+        } else if e.has_epoch_not_match() {
+            Self::on_region_epoch_not_match(
+                pd_client.clone(),
+                region_store,
+                e.take_epoch_not_match(),
+            )
+            .await
+        } else if e.has_stale_command() || e.has_region_not_found() {
+            pd_client.invalidate_region_cache(ver_id).await;
+            Ok(false)
+        } else if e.has_server_is_busy()
+            || e.has_raft_entry_too_large()
+            || e.has_max_timestamp_not_synced()
+        {
+            Err(Error::RegionError(e))
+        } else {
+            info!("unknwon region error: {:?}", e);
+            pd_client.invalidate_region_cache(ver_id).await;
+            Ok(false)
+        }
+    }
+
+    // Returns
+    // 1. Ok(true): error has been resolved, retry immediately
+    // 2. Ok(false): backoff, and then retry
+    // 3. Err(Error): can't be resolved, return the error to upper level
+    async fn on_region_epoch_not_match(
+        pd_client: Arc<PdC>,
+        region_store: RegionStore,
+        error: EpochNotMatch,
+    ) -> Result<bool> {
+        let ver_id = region_store.region_with_leader.ver_id();
+        if error.get_current_regions().is_empty() {
+            pd_client.invalidate_region_cache(ver_id).await;
+            return Ok(true);
+        }
+
+        for r in error.get_current_regions() {
+            if r.get_id() == region_store.region_with_leader.id() {
```

Comment:
It doesn't insert new regions to the cache?

Reply:
The current error handling is coarse-grained. I'd like to leave these improvements to following PRs to keep this one relatively simple.